### PR TITLE
Grading Discussions

### PR DESCRIPTION
**Pull request by andrey-canon**
## Description 
This allows to grade contributions to Rocket.Chat, reward students who are engaged with the discussion.
https://docs.google.com/document/d/1Cd-wuyezWRbaDKFfi_5v2RN59hCUqMYZ9r8HkTCHsiM/edit

## Reviewers 
- [x] @jfavellar90 
- [ ] @diegomillan
- [ ] @nicovanniekerk

### CODE REVIEW

File: rocketc/rocketc.py [hunk 1]
```diff
@@ -71,12 +71,68 @@ class RocketChatXBlock(XBlock, XBlockWithSettingsMixin, StudioEditableXBlockMixi
...
+        default="",
+        scope=Scope.settings,
+        help="Select the emoji which you want to grade",
+    )
+
+    oldest = DateTime(
+        display_name="Date From",
+        default=None,
+        scope=Scope.settings,
+        help="ISO-8601 formatted string representing the start date of this assignment."
```
Comment (jfavellar90) [inline rocketc/rocketc.py]: Can we add an example of this date format in brackets?

File: rocketc/rocketc.py [hunk 2]
```diff
@@ -71,12 +71,68 @@ class RocketChatXBlock(XBlock, XBlockWithSettingsMixin, StudioEditableXBlockMixi
         scope=Scope.user_state
     )
 
+    emoji = String(
+        display_name="Emoji to grade with",
+        default="",
+        scope=Scope.settings,
+        help="Select the emoji which you want to grade",
```
Comment (jfavellar90) [inline rocketc/rocketc.py]: is this an input to enter a string representing a emoji?

File: rocketc/rocketc.py [hunk 3]
```diff
@@ -71,12 +71,68 @@ class RocketChatXBlock(XBlock, XBlockWithSettingsMixin, StudioEditableXBlockMixi
...
+        default=None,
+        scope=Scope.settings,
+        help="ISO-8601 formatted string representing the start date of this assignment."
+    )
+
+    latest = DateTime(
+        display_name="Date To",
+        default=None,
+        scope=Scope.settings,
+        help="ISO-8601 formatted string representing the due date of this assignment."
```
Comment (jfavellar90) [inline rocketc/rocketc.py]: Example

File: rocketc/rocketc.py [hunk 4]
```diff
@@ -71,12 +71,68 @@ class RocketChatXBlock(XBlock, XBlockWithSettingsMixin, StudioEditableXBlockMixi
...
+        default=False,
+        scope=Scope.settings,
+    )
+
+    weight = Float(
+        display_name="Score",
+        help="Defines the number of points each problem is worth. ",
+        values={"min": 0, "step": .1},
+        default=1,
+        scope=Scope.settings
```
Comment (jfavellar90) [inline rocketc/rocketc.py]: what is a problem in this rocket chat context?

File: rocketc/rocketc.py [hunk 5]
```diff
@@ -71,12 +71,68 @@ class RocketChatXBlock(XBlock, XBlockWithSettingsMixin, StudioEditableXBlockMixi
...
+    grade = Float(
+        scope=Scope.user_state,
+        default=0
+    )
+
+    count_messages = Integer(
+        display_name="Last Messages",
+        default=1000,
+        scope=Scope.settings,
+        help="The amount of messages to retrieve"
```
Comment (jfavellar90) [inline rocketc/rocketc.py]: What is this value for? to retrieve the last n messages of a group/channel?

File: rocketc/rocketc.py [hunk 6]
```diff
@@ -538,3 +596,72 @@ def _get_list_groups(self, user_id, auth_token, query=""):
...
+            return None
+
+        return [message for message in messages["messages"]
+                if message["u"]["username"] == self.user_data["username"]]
+
+    def _filter_by_reaction_and_user_role(self, messages, reaction):
+        """
+        Returns generator with filtered messages by a given reaction
+        """
+        for message in messages:
```
Comment (jfavellar90) [inline rocketc/rocketc.py]: what about using the filter function? https://docs.python.org/2/library/functions.html#filter

File: rocketc/rocketc.py [hunk 7]
```diff
@@ -538,3 +596,72 @@ def _get_list_groups(self, user_id, auth_token, query=""):
...
+            return True
+        return False
+
+    def _grading_discussions(self):
+        """
+        This method allows to grade contributions to Rocket.Chat given a reaction.
+        """
+        if not self.graded_activity or self.grade == self.weight:
+            return
+        messages = self._get_user_messages(self.user_data["default_group"],
```
Comment (jfavellar90) [inline rocketc/rocketc.py]: Only graded on the default group?

File: rocketc/rocketc.py [hunk 8]
```diff
@@ -264,6 +320,8 @@ def init(self):
 
             if user_data["role"] == "instructor" and self.rocket_chat_role != "bot":
                 self.api_rocket_chat.change_user_role(user_id, "bot")
+
```
Comment (jfavellar90) [inline rocketc/rocketc.py]: everytime the xblock loads for a user, the grade is calculated? Could it cause performance issues?

File: rocketc/rocketc.py [hunk 9]
```diff
@@ -538,3 +596,72 @@ def _get_list_groups(self, user_id, auth_token, query=""):
...
+        This method allows to grade contributions to Rocket.Chat given a reaction.
+        """
+        if not self.graded_activity or self.grade == self.weight:
+            return
+        messages = self._get_user_messages(self.user_data["default_group"],
+                                           self.latest, self.oldest, self.count_messages)
+        messages = list(self._filter_by_reaction_and_user_role(messages, self.emoji))
+        if len(messages) >= self.target_reaction:
+            self.grade = self.weight
+            self.runtime.publish(self, 'grade', {'value': self.grade, 'max_value': self.weight})
```
Comment (jfavellar90) [inline rocketc/rocketc.py]: What does this line do exactly?

Comment (jfavellar90) [reply rocketc/rocketc.py]: self.runtime.publish

File: rocketc/rocketc.py (hunk 5 above)
Comment (jfavellar90) [reply rocketc/rocketc.py]: Yes (answering myself)

File: rocketc/rocketc.py (hunk 7 above)
Comment (andrey-canon) [reply rocketc/rocketc.py]: Yes, because the user can be in a lot of groups

File: rocketc/rocketc.py (hunk 9 above)
Comment (andrey-canon) [reply rocketc/rocketc.py]: Publish the grade

File: rocketc/rocketc.py (hunk 6 above)
Comment (andrey-canon) [reply rocketc/rocketc.py]: What is the advantage ?

File: rocketc/rocketc.py (hunk 4 above)
Comment (andrey-canon) [reply rocketc/rocketc.py]: The problem is get a number of emojis for a specific comment on rocketchat

File: rocketc/rocketc.py (hunk 1 above)
Comment (jfavellar90) [reply rocketc/rocketc.py]: Good

File: rocketc/rocketc.py (hunk 9 above)
Comment (jfavellar90) [reply rocketc/rocketc.py]: Where?

File: rocketc/rocketc.py (hunk 6 above)
Comment (jfavellar90) [reply rocketc/rocketc.py]: For me is more legible, but according to this post, you can use also list comprehension: https://stackoverflow.com/questions/3013449/list-comprehension-vs-lambda-filter/3013686

File: rocketc/rocketc.py (hunk 9 above)
Comment (andrey-canon) [reply rocketc/rocketc.py]: https://github.com/edx/edx-documentation/blob/master/en_us/xblock-tutorial/source/concepts/events.rst#id11

File: rocketc/rocketc.py (hunk 6 above)
Comment (jfavellar90) [reply rocketc/rocketc.py]: Ok after our discussion, just reduce the the two ifs to just one

**Comment by andrey-canon** [comment]
@jfavellar90, your suggestions are ready in the last commit

**Pull request merged**